arroyo-connectors/src/impulse.rs

```diff
@@ -0,0 +1,178 @@
+            source_field("counter", Primitive(UInt64 as i32)),
+            source_field("subtask_index", Primitive(UInt64 as i32)),
+        ],
+        format_options: None,
+        definition: None,
+    }
+}
+
+pub struct ImpulseConnector {}
+
+impl Connector for ImpulseConnector {
+    type ConfigT = EmptyConfig;
+    type TableT = ImpulseTable;
+
+    fn name(&self) -> &'static str {
+        "impulse"
+    }
+
+    fn metadata(&self) -> grpc::api::Connector {
+        grpc::api::Connector {
+            id: "impulse".to_string(),
+            name: "Impulse".to_string(),
+            icon: ICON.to_string(),
+            description: "Periodic demo source".to_string(),
+            enabled: true,
+            source: true,
+            sink: false,
+            testing: false,
+            custom_schemas: false,
+            connection_config: None,
+            table_config: TABLE_SCHEMA.to_string(),
+        }
+    }
+
+    fn table_type(&self, _: Self::ConfigT, _: Self::TableT) -> arroyo_rpc::grpc::api::TableType {
+        return grpc::api::TableType::Source;
+    }
+
+    fn get_schema(
+        &self,
+        _: Self::ConfigT,
+        _: Self::TableT,
+        _: Option<&ConnectionSchema>,
+    ) -> Option<ConnectionSchema> {
+        Some(impulse_schema())
+    }
+
+    fn test(
+        &self,
+        _: &str,
+        _: Self::ConfigT,
+        _: Self::TableT,
+        _: Option<&arroyo_rpc::grpc::api::ConnectionSchema>,
+        tx: tokio::sync::mpsc::Sender<
+            Result<arroyo_rpc::grpc::api::TestSourceMessage, tonic::Status>,
+        >,
+    ) {
+        tokio::task::spawn(async move {
+            tx.send(Ok(TestSourceMessage {
+                error: false,
+                done: true,
+                message: "Successfully validated connection".to_string(),
+            }))
+            .await
+            .unwrap();
+        });
+    }
+
+    fn from_options(
+        &self,
+        name: &str,
+        options: &mut std::collections::HashMap<String, String>,
+        s: Option<&arroyo_rpc::grpc::api::ConnectionSchema>,
+    ) -> anyhow::Result<Connection> {
+        let event_rate = f64::from_str(&pull_opt("event_rate", options)?)
+            .map_err(|_| anyhow!("invalid value for event_rate; expected float"))?;
+
+        let event_time_interval: Option<i64> = options
+            .remove("event_time_interval")
+            .map(|t| i64::from_str(&t))
+            .transpose()
+            .map_err(|_| anyhow!("invalid value for event_time_interval; expected float"))?;
+
+        let message_count: Option<i64> = options
+            .remove("message_count")
+            .map(|t| i64::from_str(&t))
+            .transpose()
+            .map_err(|_| anyhow!("invalid value for event_time_interval; expected float"))?;
+
+        // validate the schema
+        if let Some(s) = s {
+            if s.fields != impulse_schema().fields {
+                bail!("invalid schema for impulse source");
```

Comment:
Might be good to return the expected schema. It's also currently very strict. The field order needs to be (counter, subtask_index), and the type has to be`bigint unsigned`.

Reply:
Yeah, it would be good, but with our current error reporting capabilities in the UI we only support one or two lines of text

I think the real answer here is CREATE EXTERNAL so users don't have to specify a schema for connectors that have a specific schema